policies/recipes/steps.go

```diff
@@ -90,8 +99,344 @@ func parsePermissions(s string) (os.FileMode, error) {
 }
 
 // StepArchiveExtraction builds the command for a ArchiveExtraction step
-func StepArchiveExtraction(step *osconfigpb.SoftwareRecipe_Step_ArchiveExtraction, artifacts map[string]string) error {
-	fmt.Println("StepArchiveExtraction")
+func StepArchiveExtraction(step *osconfigpb.SoftwareRecipe_Step_ArchiveExtraction, artifacts map[string]string, stepDir string) error {
+	filename, ok := artifacts[step.ArchiveExtraction.GetArtifactId()]
+	if !ok {
+		return fmt.Errorf("%q not found in artifact map", step.ArchiveExtraction.GetArtifactId())
+	}
+	switch step.ArchiveExtraction.GetType() {
+	case osconfigpb.SoftwareRecipe_Step_ExtractArchive_ZIP:
+		return extractZip(filename, step.ArchiveExtraction.Destination)
+	case osconfigpb.SoftwareRecipe_Step_ExtractArchive_TAR:
+		return extractTar(filename, step.ArchiveExtraction.Destination)
+	case osconfigpb.SoftwareRecipe_Step_ExtractArchive_TAR_GZIP:
+		compressed, err := os.Open(filename)
```

Comment:
Don't decompress to a file, just pass the reader along to tar. Also all these opens should happen before the switch, then extractZip and extractTar should just take a io.Reader. something like:

file, err := os.Open(filename)
defer file.Close()
step.ArchiveExtraction.GetType() {
case osconfigpb.SoftwareRecipe_Step_ExtractArchive_ZIP:
return extractZip(file, step.ArchiveExtraction.Destination)
case osconfigpb.SoftwareRecipe_Step_ExtractArchive_TAR:
return extractTar(file, step.ArchiveExtraction.Destination)
case osconfigpb.SoftwareRecipe_Step_ExtractArchive_TAR_GZIP:
gz, err := gzip.NewReader(file)
defer reader.Close()
return extractTar(gz, step.ArchiveExtraction.Destination)

Reply:
So the issue here is that we go through the file twice so it seemed like we either need to decompress it temporarily or decompress it twice (I'm not sure which of those is practically better).

Also we could switch it to only one loop and not check for conflicts early and the we would need to only use the reader once if that would make sense but I'm unsure whether this is better or not.

Reply:
I'm not sure I understand, a file allows for seeking and I would argue CPU cycles are cheaper (and faster) than disk. Also we aren't extracting twice, the seek is faster than the full extract, also efficiency is already really poor as we run an additional stat for every file to check for conflicts.

Reply:
I don't understand how we would seek a gzip reader, in the tar.gz case

Reply:
I switched it to not extracting to a file. I had to reinitialize the gzip and tar readers after seeking the files